packages/actor-query-operation-leftjoin-nestedloop/lib/ActorQueryOperationLeftJoinNestedLoop.ts

```diff
@@ -20,37 +24,71 @@ export class ActorQueryOperationLeftJoinNestedLoop extends ActorQueryOperationTy
-        rightStream.on('data', (rightItem) => {
-          const join = ActorRdfJoin.join(leftItem, rightItem);
-          if (join) {
-            // if (pattern.expression) {
-            //   // TODO: do this once expressions are implemented
-            //   // Values of both streams can be used for this expression
-            // }
-
-            bindingsStream._push(join);
-          }
+    const leftRaw = await this.mediatorQueryOperation.mediate({ operation: pattern.left, context });
+    const left = ActorQueryOperation.getSafeBindings(leftRaw);
+    const rightRaw = await this.mediatorQueryOperation.mediate({ operation: pattern.right, context });
+    const right = ActorQueryOperation.getSafeBindings(rightRaw);
+
+    const leftJoinInner = (outerItem: Bindings, innerStream: ClonedIterator<Bindings>) => {
+      const joinedStream = innerStream
+        .transform({
+          transform: async (innerItem: Bindings, nextInner: any) => {
+            const joinedBindings = ActorRdfJoin.join(outerItem, innerItem);
+            if (!joinedBindings) { nextInner(); return; }
+            if (!pattern.expression) {
+              joinedStream._push({ joinedBindings, result: true });
+              nextInner();
+              return;
+            }
+            try {
+              const evaluator = new AsyncEvaluator(pattern.expression);
+              const result = await evaluator.evaluateAsEBV(joinedBindings);
+              joinedStream._push({ joinedBindings, result });
+            } catch (err) {
+              if (!this.isExpressionError(err)) {
+                bindingsStream.emit('error', err);
+              }
+            }
+            nextInner();
+          },
         });
-      },
-    });
+      return joinedStream;
+    };
+
+    const leftJoinOuter = (leftItem: Bindings, nextLeft: any) => {
+      const innerStream = right.bindingsStream.clone(); // TODO: Why does this even work?
+      const joinedStream = leftJoinInner(leftItem, innerStream);
+
+      // TODO: Does this even work for large streams?
+      // The full inner stream is kept in memory.
+      joinedStream.on('end', () => nextLeft());
+      joinedStream.on('data', async ({ joinedBindings, result }) => {
+        if (result) {
+          bindingsStream._push(joinedBindings);
+        }
+      });
+    };
+
+    const transform = leftJoinOuter;
+    const bindingsStream = left.bindingsStream
+      .transform<Bindings>({ optional: true, transform });
 
     const variables = ActorRdfJoin.joinVariables({ entries: [left, right] });
     const metadata = () => Promise.all([pattern.left, pattern.right].map((entry) => entry.metadata))
       .then((metadatas) => metadatas.reduce((acc, val) => acc * val.totalItems, 1))
       .catch(() => Infinity)
-      .then((totalItems) =>  ({ totalItems }) );
+      .then((totalItems) => ({ totalItems }));
 
     return { type: 'bindings', bindingsStream, metadata, variables };
   }
 
+  // TODO Duplication with e.g. Extend
```

Comment:
Maybe we can move this function to sparqlee?

Reply:
The only reason it's a separate function is so that it can be easily mocked for tests, I'll have to check if that's still a an option. Another elegant solution would be to just provide this on the query operation abstract class, this way the documentation stays in house too.

Reply:
> The only reason it's a separate function is so that it can be easily mocked for tests

It should be possible to mock specific functions from (external) modules with Jest.

> Another elegant solution would be to just provide this on the query operation abstract class

I would create a separate package for this then, as this would require a dependency on sparqlee, while this is only needed for some query operators.